src/garage/_environment.py

```
@@ -0,0 +1,226 @@
+
+    Example of a simple rollout loop:
+
+    .. code-block:: python
+
+        env = MyEnv()
+        policy = MyPolicy()
+        first_ts = env.reset()
+        env.visualize()  # visualization window opened
+
+        episode = [first_ts, ]
+        while not ts.last():
+            action = policy.get_action(episode[-1].next_observation)
+            episode.append(env.step(action))
+
+        env.close()  # visualization window closed
+
+    Make sure your environment is pickle-able:
+        Garage pickles the environment via the `cloudpickle` module
+        to save snapshots of the experiment. However, some environments may
+        contain attributes that are not pickle-able (e.g. a client-server
+        connection). In such cases, override `__setstate__()` and
+        `__getstate()__` to add your custom pickle logic.
+
+        You might want to refer to the EzPickle module:
+        https://github.com/openai/gym/blob/master/gym/utils/ezpickle.py
+        for a lightweight way of pickle and unpickle via constructor
+        arguments.
+
+    """
+
+    @property
+    @abc.abstractmethod
+    def action_space(self):
+        """Return the action space specification.
+
+        Some environments do not provide their action space specifications as
+        type `akro.Space`. In such cases, subclasses should silently convert
+        the environment's action_space to `akro.Space`.
+
+        Returns:
+            akro.Space: The environment specification.
+
+        """
+
+    @property
+    @abc.abstractmethod
+    def observation_space(self):
+        """Return the observation space specification.
+
+        Some environments do not provide their observation space
+        specifications as type `akro.Space`. In such cases, subclasses should
+        silently convert the environment's observation_space to `akro.Space`.
+
+        Returns:
+            akro.Space: The environment specification.
```

this isn't an env spec?

---

The observation space is akro.Space. And spec is EnvSpec?

---

...but the `observation_space` is not `The environment specification`?

By the way, I don't think you need to say anything about "silently convert" -- this is the case with all APIs.

---

Ah I see. Will change to observation space specification.